In [9]:
import pandas as pd
import matplotlib.pyplot as plt
import json
import tensorflow as tf
from tensorflow import keras
from transformers import BertTokenizer, TFBertModel

In [10]:
def parse_jsonl(path: str) -> pd.DataFrame:
    data = {}
    with open(path, 'r') as f:
        for (i, line) in enumerate(f):
            line = json.loads(line)
            line["conversation_id"] = [i] * len(line["messages"])
            for k in line.keys():
                if k == "players":
                    continue
                if k == "game_id":
                    line[k] = [line[k]] * len(line["messages"])
                data[k] = data.get(k, []) + line[k]
    return pd.DataFrame(data)

In [11]:
def clean_data(df: pd.DataFrame) -> pd.DataFrame:
    df.loc[df["receiver_labels"] == "NOANNOTATION", "receiver_labels"] = pd.NA
    df["receiver_labels"] = df["receiver_labels"].astype(bool)
    return df

In [12]:
def pipe(*funcs):
    out = funcs[0]
    for func in funcs[1:]:
        out = func(out)
    return out

In [13]:
def docs_to_vocab_ids(tokenized_texts_list):
    """
    converting a list of strings to a list of lists of word ids
    """
    texts_vocab_ids = []
    text_labels = []
    valid_example_list = []
    for i, token_list in enumerate(tokenized_texts_list):

        # Get the vocab id for each token in this doc ([UNK] if not in vocab)
        vocab_ids = []
        for token in list(token_list.numpy()):
            decoded = token.decode('utf-8', errors='ignore')
            if decoded in vocab_dict:
                vocab_ids.append(vocab_dict[decoded])
            else:
                vocab_ids.append(vocab_dict['[UNK]'])

        # Truncate text to max length, add padding up to max length
        vocab_ids = vocab_ids[:MAX_SEQUENCE_LENGTH]
        n_padding = (MAX_SEQUENCE_LENGTH - len(vocab_ids))
        # For simplicity in this model, we'll just pad with unknown tokens
        vocab_ids += [vocab_dict['[UNK]']] * n_padding
        # Add this example to the list of converted docs
        texts_vocab_ids.append(vocab_ids)

        if i % 5000 == 0:
            print('Examples processed: ', i)

    print('Total examples: ', i)
    return np.array(texts_vocab_ids)

In [14]:
MAX_SEQUENCE_LENGTH = 10000

In [15]:
bert_tokenizer = BertTokenizer.from_pretrained('bert-base-cased')
bert_model = TFBertModel.from_pretrained('bert-base-cased')
# train_encodings = bert_tokenizer(train_texts, truncation=True, padding=True, max_length=max_length, return_tensors='tf')
# valid_encodings = bert_tokenizer(valid_texts, truncation=True, padding=True, max_length=max_length, return_tensors='tf')
# test_encodings = bert_tokenizer(test_texts, truncation=True, padding=True, max_length=max_length, return_tensors='tf')

Some weights of the PyTorch model were not used when initializing the TF 2.0 model TFBertModel: ['cls.predictions.bias', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.dense.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing TFBertModel from a PyTorch model trained on another task or with another architecture (e.g. initializing a TFBertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing TFBertModel from a PyTorch model that you expect to be exactly identical (e.g. initializing a TFBertForSequenceClassification model from a BertForSequenceClassification model).
All the weights of TFBertModel were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFBertModel for predictions w

# data explortion

In [16]:
train_df = pd.read_json("../data/train.jsonl", lines=True)
train_df

messages  \
0    [Germany!\n\nJust the person I want to speak w...   
1    [Hello there! What's your general plan for thi...   
2    [Buongiorno! \nBe kinda nice to know if you're...   
3    [Hey italy! good luck this game. I'm guessing ...   
4    [Hello Italy what’s up what are your thoughts ...   
..                                                 ...   
184  [Greetings Sultan!\n\nAs your neighbor I would...   
185  [Greetings My Good Frenchman! \n\nHow are your...   
186                                       [Hey, Hello]   
187  [Hello France, world you like to discuss worki...   
188                                                 []   

                                         sender_labels  \
0    [True, True, True, True, True, True, True, Tru...   
1    [True, False, True, False, True, True, True, T...   
2    [True, True, False, True, True, True, True, Tr...   
3    [True, False, True, True, True, True, True, Tr...   
4    [True, False, False, True, True, True, True, T...   
..                                                 ...   
184  [False, True, False, True, True, True, True, T...   
185  [True, True, True, True, True, True, True, Fal...   
186                                       [True, True]   
187  [False, True, True, True, True, True, True, Tr...   
188                                                 []   

                                       receiver_labels  \
0    [True, True, True, True, NOANNOTATION, NOANNOT...   
1    [True, True, True, True, True, NOANNOTATION, T...   
2    [True, False, True, False, True, True, NOANNOT...   
3    [NOANNOTATION, True, True, False, True, True, ...   
4    [NOANNOTATION, True, True, True, True, True, N...   
..                                                 ...   
184  [True, True, True, True, True, True, True, Tru...   
185  [True, True, False, True, True, True, True, Tr...   
186                                       [True, True]   
187  [True, True, True, True, True, True, False, Tr...   
188                                                 []   

                                              speakers  \
0    [italy, germany, italy, germany, italy, italy,...   
1    [austria, italy, austria, italy, italy, austri...   
2    [russia, italy, russia, italy, russia, italy, ...   
3    [england, italy, england, england, england, it...   
4    [turkey, italy, italy, italy, turkey, italy, t...   
..                                                 ...   
184  [russia, turkey, russia, russia, russia, turke...   
185  [russia, france, russia, russia, france, franc...   
186                                  [england, turkey]   
187  [england, france, england, england, england, f...   
188                                                 []   

                                             receivers  \
0    [germany, italy, germany, italy, germany, germ...   
1    [italy, austria, italy, austria, austria, ital...   
2    [italy, russia, italy, russia, italy, russia, ...   
3    [italy, england, italy, italy, italy, england,...   
4    [italy, turkey, turkey, turkey, italy, turkey,...   
..                                                 ...   
184  [turkey, russia, turkey, turkey, turkey, russi...   
185  [france, russia, france, france, russia, russi...   
186                                  [turkey, england]   
187  [france, england, france, france, france, engl...   
188                                                 []   

                                absolute_message_index  \
0    [74, 76, 86, 87, 89, 92, 97, 117, 119, 121, 12...   
1    [1, 67, 71, 73, 98, 99, 101, 179, 181, 185, 18...   
2    [11, 50, 52, 57, 61, 66, 77, 85, 96, 102, 116,...   
3    [32, 95, 106, 107, 108, 110, 113, 125, 126, 12...   
4    [45, 94, 103, 150, 154, 178, 192, 194, 195, 19...   
..                                                 ...   
184  [78, 107, 145, 370, 371, 374, 415, 420, 495, 4...   
185  [75, 115, 147, 176, 205, 206, 254, 285, 306, 5...   
186                                    

In [17]:
train_df = (parse_jsonl("../data/train.jsonl"))
test_df = (parse_jsonl("../data/test.jsonl"))
validation_df = (parse_jsonl("../data/validation.jsonl"))

In [20]:
train_df

messages  sender_labels  \
0      Germany!\n\nJust the person I want to speak wi...           True   
1      You've whet my appetite, Italy. What's the sug...           True   
2                                                      👍           True   
3      It seems like there are a lot of ways that cou...           True   
4      Yeah, I can’t say I’ve tried it and it works, ...           True   
...                                                  ...            ...   
13127  Is there any way of me actually ending this co...           True   
13128      Can we agree on peace? What are your demands?           True   
13129  Neutrality in exchange for current holdings, S...           True   
13130  Thats a bit too much, can I keep Spain and i h...           True   
13131                                         Any deals?           True   

      receiver_labels speakers receivers  absolute_message_index  \
0                True    italy   germany                      74   
1                True  germany     italy                      76   
2                True    italy   germany                      86   
3                True  germany     italy                      87   
4        NOANNOTATION    italy   germany                      89   
...               ...      ...       ...                     ...   
13127            True   france   england                     380   
13128            True   france   england                     433   
13129           False  england    france                     434   
13130            True   france   england                     437   
13131            True   france   england                     467   

       relative_message_index seasons years game_score game_score_delta  \
0                           0  Spring  1901          3                0   
1                           1  Spring  1901          3                0   
2                           2  Spring  1901          3                0   
3                           3  Spring  1901          3                0   
4                           4  Spring  1901          3                0   
...                       ...     ...   ...        ...              ...   
13127                      11    Fall  1902          4               -1   
13128                      12    Fall  1902          4               -1   
13129                      13    Fall  1902          5                1   
13130                      14    Fall  1902          4               -1   
13131                      15  Spring  1903          3               -3   

       game_id  conversation_id  
0            1                0  
1            1                0  
2            1                0  
3            1                0  
4            1                0  
...        ...              ...  
13127       10              187  
13128       10              187  
13129       10              187  
13130       10              187  
13131       10              187  

[13132 rows x 13 columns]

We can see that there are some `receiver_labels` that have the value `NOANNOTATION`. this is not a bool value 

In [16]:
round(len(train_df[train_df["receiver_labels"] == "NOANNOTATION"].index) /  len(train_df.index), 4) * 100

8.43

About 8.43% of the rows have `NOANNOTATION`

It is a small amount that we would removed, but the process may not use the `receiver_labels` columns.

The columns that may be used is `messages` and `sender_labels`

# Test 1

In [ ]:
def create_bert_multiclass_model(checkpoint = 'bert-base-cased',
                                 num_classes = 20,
                                 hidden_size = 201,
                                 dropout=0.3,
                                 learning_rate=0.00005):
    """
    Build a simple classification model with BERT. Use the Pooler Output for classification purposes.
    """
    input_ids = tf.keras.layers.Input(shape=(MAX_SEQUENCE_LENGTH,), dtype=tf.int64, name='input_ids_layer')
    token_type_ids = tf.keras.layers.Input(shape=(MAX_SEQUENCE_LENGTH,), dtype=tf.int64, name='token_type_ids_layer')
    attention_mask = tf.keras.layers.Input(shape=(MAX_SEQUENCE_LENGTH,), dtype=tf.int64, name='attention_mask_layer')

    classification_output = pipe(
        {
            'input_ids': input_ids,
            'token_type_ids': token_type_ids,
            'attention_mask': attention_mask
        },                                                                              # input
        TFBertModel.from_pretrained(checkpoint, trainable=True),                        # bert model
        lambda x: x[1],                                                                 # pooler output
        tf.keras.layers.Dense(hidden_size, activation='relu', name='hidden_layer'),     # dense layer
        tf.keras.layers.Dropout(dropout, name="dropout_layer"),                         # dropout layer
        tf.keras.layers.Dense(num_classes, activation='softmax', name='ouput_layer')    # ouput layer
    )

    classification_model = tf.keras.Model(inputs=[input_ids, token_type_ids, attention_mask], outputs=classification_output)

    # Compile the model with categorical cross-entropy loss and an optimizer
    classification_model.compile(
        loss='sparse_categorical_crossentropy',
        optimizer=tf.keras.optimizers.Adam(learning_rate=learning_rate),
        metrics=['accuracy']
    )

    return classification_model

In [ ]:
def create_bert_cls_model(bert_base_model,
                          max_sequence_length=MAX_SEQUENCE_LENGTH,
                          hidden_size = 100,
                          dropout=0.3,
                          learning_rate=0.00005):
    """
    Build a simple classification model with BERT. Use the CLS Token output for classification purposes.
    """
    bert_base_model.trainable = True

    input_ids = tf.keras.layers.Input(shape=(max_sequence_length,), dtype=tf.int64, name='input_ids_layer')
    token_type_ids = tf.keras.layers.Input(shape=(max_sequence_length,), dtype=tf.int64, name='token_type_ids_layer')
    attention_mask = tf.keras.layers.Input(shape=(max_sequence_length,), dtype=tf.int64, name='attention_mask_layer')

    bert_inputs = {'input_ids': input_ids,
                   'token_type_ids': token_type_ids,
                   'attention_mask': attention_mask}

    classification_ouput = pipe(
        bert_model(bert_inputs)[0][:, 0, :],
        tf.keras.layers.Dense(hidden_size, activation='relu', name='hidden_layer'),
        tf.keras.layers.Dropout(dropout),
        tf.keras.layers.Dense(1, activation='sigmoid', name='classification_layer')
    )

    classification_model = tf.keras.Model(inputs=[input_ids, token_type_ids, attention_mask], outputs=[classification_ouput])

    # Compile the model
    classification_model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=learning_rate),
                    loss='binary_crossentropy',
                    metrics=['accuracy'])

    return classification_model

# TEST 2

In [ ]:
def create_lstm_model():
    model = keras.Sequential([
        keras.layers.Embedding()
    ])